static/js/components/Spectrum.jsx

```diff
@@ -0,0 +1,243 @@
+import React, { useEffect } from "react";
+import { useDispatch, useSelector } from "react-redux";
+import PropTypes from "prop-types";
+import Grid from "@material-ui/core/Grid";
+import Typography from "@material-ui/core/Typography";
+import Paper from "@material-ui/core/Paper";
+import Button from "@material-ui/core/Button";
+import Dialog from "@material-ui/core/Dialog";
+import DialogContent from "@material-ui/core/DialogContent";
+import { Link } from "react-router-dom";
+import DownloadLink from "react-download-link";
+import { makeStyles } from "@material-ui/core/styles";
+import dayjs from "dayjs";
+import Papa from "papaparse";
+
+import { showNotification } from "baselayer/components/Notifications";
+import Plot from "./Plot";
+import { UserContactInfo } from "./UserProfileInfo";
+import { fetchSourceSpectra, deleteSpectrum } from "../ducks/spectra";
+
+const useStyles = makeStyles({
+  plot: {
+    width: "900px",
+    overflow: "auto",
+  },
+  inner: { padding: "1rem" },
+  margined: { margin: "1rem" },
+  detailedSpecButton: {
+    color: "primary",
+    variant: "contained",
+    cursor: "default",
+  },
+  specTitle: { margin: "0.5rem" },
+});
+
+function get_filename(spectrum, instrument) {
+  return `${spectrum.obj_id}_${instrument.name}_${spectrum.observed_at}.csv`;
+}
+
+function to_csv(spectrum) {
+  const formatted = [];
+  spectrum.wavelengths.forEach((wave, i) => {
+    const obj = {};
+    obj.wavelength = wave;
+    obj.flux = spectrum.fluxes[i];
+    if (spectrum.fluxerr) {
+      obj.fluxerr = spectrum.fluxerr[i];
+    }
+    formatted.push(obj);
+  });
+  return Papa.unparse(formatted);
+}
+
+const DetailedSpectrumView = ({ spectrum }) => {
+  const classes = useStyles();
+  const dispatch = useDispatch();
+  const { instrumentList } = useSelector((state) => state.instruments);
+  const { id: uid } = useSelector((state) => state.profile);
+  const [open, setOpen] = React.useState(false);
+
+  const instrument = instrumentList.find(
+    (i) => i.id === spectrum.instrument_id
+  );
+
+  const data = spectrum.original_file_string
+    ? spectrum.original_file_string
+    : to_csv(spectrum);
+  const filename = spectrum.original_file_filename
+    ? spectrum.original_file_filename
+    : get_filename(spectrum, instrument);
+
+  return (
+    <div>
+      <Typography variant="h6">Uploaded by</Typography>
+      <UserContactInfo user={spectrum.owner} />
+      {spectrum.reducers.length > 0 && (
```

Comment:
```suggestion
      {spectrum.reducers?.length > 0 && (
```
Otherwise it'll error if reducers is absent. Same for observers, etc, below

Reply:
done